apps/passport-server/test/database.spec.ts

```diff
@@ -207,34 +209,258 @@ describe("database reads and writes", function () {
     expect(initialStorage).to.be.undefined;
 
     const value = "value";
-    await insertEncryptedStorage(db, key, value);
+    await setEncryptedStorage(db, key, value);
     const insertedStorage = await fetchEncryptedStorage(db, key);
     if (!insertedStorage) {
       throw new Error("expected to be able to fetch a e2ee blob");
     }
     expect(insertedStorage.blob_key).to.eq(key);
     expect(insertedStorage.encrypted_blob).to.eq(value);
+    expect(insertedStorage.revision).to.eq("1");
 
     const updatedValue = "value2";
     expect(value).to.not.eq(updatedValue);
-    await insertEncryptedStorage(db, key, updatedValue);
+    await setEncryptedStorage(db, key, updatedValue);
     const updatedStorage = await fetchEncryptedStorage(db, key);
     if (!updatedStorage) {
       throw new Error("expected to be able to fetch updated e2ee blog");
     }
     expect(updatedStorage.blob_key).to.eq(key);
     expect(updatedStorage.encrypted_blob).to.eq(updatedValue);
+    expect(updatedStorage.revision).to.eq("2");
 
     // Storing an empty string is valid, and not treated as deletion.
     const emptyValue = "";
     expect(updatedValue).to.not.eq(emptyValue);
-    await insertEncryptedStorage(db, key, emptyValue);
+    await setEncryptedStorage(db, key, emptyValue);
     const emptyStorage = await fetchEncryptedStorage(db, key);
     if (!emptyStorage) {
       throw new Error("expected to be able to fetch updated e2ee blog");
     }
     expect(emptyStorage.blob_key).to.eq(key);
     expect(emptyStorage.encrypted_blob).to.eq(emptyValue);
+    expect(emptyStorage.revision).to.eq("3");
+  });
+
+  step("e2ee update conflict detection should work", async function () {
+    const key = "ckey1";
+    const initialStorage = await fetchEncryptedStorage(db, key);
+    expect(initialStorage).to.be.undefined;
+
+    // Can't use "update" to set initial state.
+    const value1 = "value1";
+    const missingResult = await updateEncryptedStorage(db, key, value1, "0");
+    expect(missingResult.status).to.eq("missing");
+    expect(missingResult.revision).to.be.undefined;
+
+    // Use "set" to create rev1.
+    const rev1 = await setEncryptedStorage(db, key, value1);
```

Comment:
one case i would insert between this one and next one is using set to create rev2

Reply:
What's the new coverage you're looking for here?  Using set() twice in a row is covered by the first test case.  You can't use set() and update() to create the same rev because revs only ever increment.  You can't set() with the wrong knownRevision because it doesn't take that argument at all.  Maybe you want interspersing of set() and update() usage?

Reply:
I was looking for ' Using set() twice in a row ' - good to know you have that covered. feel free to ignore.

Reply:
Yeah, I feel like there's a lot of duplication of the with-rev and without-rev tests, but I didn't think of a better way of doing it.  I'd like to eventually just get rid of the without-rev options when the clients can all deal with it.  The only special case it's needed for is initial creation, and I'm on the fence about whether that deserves to be a special case "create" function so it can be conflict-free too.